Autopilot/PathManager/Inc/waypointManager.hpp

```diff
@@ -0,0 +1,175 @@
+    long double longitude;            // Longitude of waypoint
+    int altitude;                     // Altitude of waypoint
+    float turnRadius;                 // If turn is commanded, this is the radius of the turn
+    char type;
+    char waypointId;                  // Id of the waypoint
+    char arrayIndex;                  // Position of the waypoint within the waypointBuffer (array)
+} _WaypointData;
+
+/*
+* Structure contains the data that will be returned to the Path Manager state manager. 
+* This data will be used by the PID and coordinated turn engine to determine the commands to be sent to the Attitude Manager.
+*/
+typedef struct {
+    uint16_t desiredHeading;            // Desired heading to stay on path
+    int desiredAltitude;                // Desired altitude at next waypoint
+    long double distanceToNextWaypoint; // Distance to the next waypoint (helps with airspeed PID)
+    _WaypointStatus errorCode;          // Contains error codes
+    bool isDataNew;                     // Notifies PID modules if the data in this structure is new
+    uint32_t timeOfData;                // The time that the data in this structure was collected
+} _WaypointManager_Data;
+
+class WaypointManager {
+    public: 
+        /*
+        * Constructor for this class
+        * 
+        * @param _WaypointData * initialWaypoints -> These waypoints will be used to initialize the waypointBuffer array
+        * @param int numWaypoints -> Number of waypoints being initialized in the waypointBuffer array
+        * @param _WaypointData* currentLocation -> Home base. 
+        */
+        WaypointManager(_WaypointData * initialWaypoints, int numWaypoints, _WaypointData* currentLocation);
+        
+        /*
+        * Updates the _WaypointManager_Data structure with new values
+        * 
+        * @param _Gps_Data currentPosition -> contains the current coordinates, altitude, and heading
+        */
+        void get_next_directions(_Gps_Data currentPosition); 
+
+        /*
+        * Adds, inserts, updates, or deletes a single waypoint in the waypointBuffer array
+        * 
+        * @param _WaypointData* waypoint -> In the instance that we need to update, insert, or append a new waypoint, this will be used 
+        * @param int waypointId -> the ID of the waypoint that will be updated or deleted
+        * @param _WaypointBufferUpdateType updateType -> the type of modification to the waypointBuffer array (look above)
+        * @param int previousId -> stores the ID of the waypoint that will come before the inserted waypoint. Set to 0 by default, so does not need to be passed
+        * @param int nextId -> stores the ID of the waypoint that will come after the inserted waypoint. Set to 0 by default, so does not need to be passed
+        */
+        void update_path_nodes(_WaypointData* waypoint, int waypointId, _WaypointBufferUpdateType updateType, int previousId = 0, int nextId = 0);
```

Comment:
Is there some way for the user to say: "Start circling here" or "continue going through the waypoint list" or "fuck it return home" ?

I guess default is circling if no waypoints are left in the list ? Or return home ?

Reply:
My idea was this:

calling void get_next_directions() basically just makes the waypoint manager go through the waypoint list as it normally would, but there are some cases where the plane will trigger certain actions:
- If the type parameter of the waypoint structure is set to 2, then the plane will undergo a holding pattern (start circling)
- If there are no more waypoints present, the plane will continue on it's current/last trajectory (we could change this to make the plane start circling) 

I could create a method that can be called to make the plane start circling/return home.